### ISSUE TITLE
Dungeon Runner: testing, QA, and bug fixes

### ISSUE THREAD

**Issue by enmaku** [enmaku/portfolio-site#55]
## Summary

Track **testing, hardening, and bug fixes** for Dungeon Runner after the Phase 2 vertical slices (#35–#44). This is the umbrella for expanding automated coverage, fixing regressions, and closing UX/engine edge cases that are not greenfield features.

## Context (from #22 / closed slices)

- Parent PRD (#22) emphasizes **TDD**, golden fixtures, legality/mask tests, player-view boundaries, persistence, and AI integration boundaries; many areas still lack **page-level or flow-level** tests.
- Phase 2 gap notes called out missing **contract tests** for `DungeonRunnerPage`, **Vorpal + Memory Aid** scenarios, **animation queue** behavior vs updated skip policy, and similar.
- Ongoing **console/runtime** hygiene, **full-match** scenarios (e.g. phases where bots must act), and **CI** confidence (lint, tests, integration checks) belong here until we split them again.

## Scope (high level — not a full spec)

- Broader and **targeted** test coverage (engine stays authoritative; add UI/flow tests where valuable).
- **Bug fixes** and small correctness patches discovered in playtesting or CI.
- **Regression** passes after larger UX or asset changes.

Detailed acceptance criteria and tickets can be broken out later.

**Comment by enmaku** [comment]
Debug note from investigation/fix:

We reproduced an intermittent freeze between dungeon runs and identified two scheduler gaps in `DungeonRunnerPage.vue`:

1) **AI turn re-entrancy/race**
- `runAiTurn()` is async (NN inference), but scheduling only guarded pending timeout IDs, not in-flight execution.
- This allowed stale/overlapping AI executions against the same logical turn window.

2) **Missing AI scheduling in `pick-adventurer`**
- After a dungeon run, phase transitions to `pick-adventurer`.
- If the picker is a bot seat, AI was never scheduled because scheduling only allowed `bidding`/`dungeon`.
- Result: game appeared frozen after “Last dungeon run”.

Fix applied:
- Added `aiTurnInFlight` guard to prevent concurrent AI execution.
- Added a stale-run token check (`matchId + turnNumber + phase + seatId`) before apply, so outdated async completions no-op.
- Extended both `runAiTurn()` and `scheduleAiTurnIfReady()` phase checks to include `pick-adventurer`.

Validation:
- Lint clean for the modified file.
- Behavior now advances past post-dungeon `pick-adventurer` instead of stalling when bot is active.

**Comment by enmaku** [comment]
Implemented and verified the dungeon-runner updates:

- New match and rematch now start in  so the first active seat chooses the first-round adventurer (no default warrior start).
- NN inference fallback no longer hard-falls to /first legal action; fallback now selects from legal actions via seeded selection, removing warrior bias on fallback hero choice.
- Updated NN runtime tests to assert legal fallback + fallback reason instead of hardcoding  expectations.

Validation run (all passing):
- ✔ chooseRandombotAction always returns a legal action (0.679916ms)
✔ chooseRandombotAction is deterministic for same state (0.443542ms)
✔ chooseRandombotAction fallback returns first legal action when total weight is non-positive (0.097834ms)
✔ chooseRandombotAction stays deterministic across mirrored state sequences (0.160625ms)
✔ createInitialMatchState is deterministic for same seed and setup (1.443417ms)
✔ createInitialMatchState includes rng progression metadata baseline (0.256375ms)
✔ createInitialMatchState enforces core seat invariants across 2-4 seats (0.246792ms)
✔ getLegalActions returns canonical actions for active seat only (0.234042ms)
✔ engine exposes python-parity phase and subphase constants (0.058333ms)
✔ pick-adventurer phase exposes choose-next-adventurer actions (0.091334ms)
✔ dungeon subphases expose python-style legal action intents (0.144292ms)
✔ vorpal declaration legality exposes full species set (0.085125ms)
✔ vorpal declaration legality ignores dungeon pile composition (0.329292ms)
✔ bidding->dungeon skips vorpal when pile non-empty but no vorpal equipment in play (0.568417ms)
✔ vorpal declaration rejects species outside legal action set (0.274709ms)
✔ bidding->dungeon transition enters vorpal subphase when dungeon has monsters (0.077958ms)
✔ bidding-to-dungeon legal actions stay timing-gated through vorpal and equipment picks (0.488333ms)
✔ dungeon decision sequence advances through python-style subphases (0.313709ms)
✔ declining fire axe does not open polymorph when no next monster exists (0.175333ms)
✔ reveal skips pick-fire-axe when B_AXE not in play (warrior center loadout) (0.16275ms)
✔ applyAction rejects invalid actions cleanly (0.0465ms)
✔ applyAction rotates turn and records rng metadata history for canonical action (0.08275ms)
✔ rng lifecycle is monotonic across successive canonical actions (0.066959ms)
✔ headless deterministic gate enforces seed sweep and invariants (2.037125ms)
✔ bidding round ends when all but one seat has passed (0.106625ms)
✔ bidding ends when all but one non-eliminated seat has passed (0.081083ms)
✔ draw action updates top card only for acting seat view (0.106083ms)
✔ sacrifice action removes selected equipment and advances turn (0.114292ms)
✔ bidding to dungeon initializes runner dungeon state like Python _end_bidding (0.177416ms)
✔ passed seats are skipped in subsequent bidding turns (0.051458ms)
✔ shuffleMatchSeats updates labels and first listed seat starts (0.1615ms)
✔ only the revealing seat can see currently revealed monster card (0.059125ms)
✔ revealed monster can be added to dungeon without sacrificing equipment (0.048875ms)
✔ runner clears dungeon via reveal chain and enters pick-adventurer (0.134833ms)
✔ runner failure in dungeon transitions to pick-adventurer with life loss (0.097625ms)
✔ match ends when a seat reaches two successes from dungeon clear (0.059ms)
✔ nn action metadata is persisted per seat for backend and fallback tracking (0.069042ms)
✔ own pile species tracking resets after picking next adventurer (0.125375ms)
✔ history row for CHOOSE_NEXT_ADVENTURER has no stale dungeonRunResult (0.081125ms)
✔ policy observation stays fixed at 87 dimensions (0.937041ms)
✔ policy legal mask maps canonical legal actions into 26-index space (0.326084ms)
✔ policy decoder supports indexed sacrifice slots (0.23525ms)
✔ policy sacrifice slot mapping follows python slot order semantics (0.110292ms)
✔ policy sacrifice slots use hero-specific loadout order (0.167625ms)
✔ policy codec maps pick-adventurer slots (0.121209ms)
✔ policy codec maps dungeon decision slots (0.164875ms)
✔ policy observation encodes bidding pending and pending-card strength (0.459708ms)
✔ policy observation encodes dungeon subphase and runner-only block (0.413625ms)
✔ policy observation uses center-equipment and full-deck normalization (1.2945ms)
✔ policy observation encodes own pile species counts (0.176583ms)
✔ policy observation and mask match exported Python fixtures (0.863917ms)
✔ nn runtime returns legal action and records backend metadata (9.742375ms)
✔ nn runtime falls back to pass when model unavailable (0.191708ms)
✔ nn runtime scheduling preserves deterministic action order (0.498291ms)
✔ nn runtime supports deterministic sampling override (0.667667ms)
✔ nn runtime serializes cross-model requests deterministically (11.295333ms)
✔ nn runtime uses worker path when available (0.308208ms)
✔ nn runtime retries once before illegal-output fallback (0.291083ms)
✔ nn runtime retries once before model-load-failed fallback (0.242833ms)
✔ nn stochastic sampling is reproducible for same state and backend (0.602541ms)
✔ nn runtime uses policy logits head and keeps dungeon reveal legal (0.308083ms)
✔ nn runtime emits debug trace payload when requested (0.20825ms)
✔ nn runtime emits debug trace on model-load fallback path (0.097166ms)
✔ worker-side model load errors do not masquerade as pass actions (0.186375ms)
ℹ tests 64
ℹ suites 0
ℹ pass 64
ℹ fail 0
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 201.926917
- 64/64 tests passed.

Files touched:
- 
- 
- 

Note: changes are currently local/uncommitted.

**Comment by enmaku** [comment]
Correction to prior comment (formatting issue). Implemented and verified the dungeon-runner updates:

- New match and rematch now start in `pick-adventurer` so the first active seat chooses the first-round adventurer (no default warrior start).
- NN inference fallback no longer hard-falls to `PASS`/first legal action; fallback now selects from legal actions via seeded selection, removing warrior bias on fallback hero choice.
- Updated NN runtime tests to assert legal fallback + fallback reason instead of hardcoding `PASS` expectations.

Validation run (all passing):
- `node --test "src/features/dungeon-runner/engine/kernel.test.js" "src/features/dungeon-runner/nn/runtime.test.js" "src/features/dungeon-runner/nn/policyAdapter.test.js" "src/features/dungeon-runner/bots/randombot.test.js"`
- 64/64 tests passed.

Files touched:
- `src/pages/projects/DungeonRunnerPage.vue`
- `src/features/dungeon-runner/nn/runtime.js`
- `src/features/dungeon-runner/nn/runtime.test.js`

Note: changes are currently local/uncommitted.
